### ISSUE TITLE
Disabled component allows keyboard interaction on links

### ISSUE THREAD

**Issue by ryelle** [WordPress/gutenberg#15799]
**Describe the bug**

In a custom block, I'm using [the `Disabled` component](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/disabled) around a `ServerSideRender`ed list, much like [the Latest Comments block](https://github.com/WordPress/gutenberg/blob/514e61298d5c056ef6ec4d3859b959dc46735dec/packages/block-library/src/latest-comments/edit.js#L92-L97), when I noticed you can tab to the links that are rendered. Since you can't mouse-click on the links, it seems like keyboard activation should also be prevented.

**To reproduce**

1. Add a Latest Comments block
2. You can't click on the rendered links
3. Tab through the toolbar to the block contents
4. You'll see your focus move from the links in the Latest Comments content
5. If you hit enter, you leave the editor and go to that link

**Expected behavior**

Perhaps [removing the href on links](https://css-tricks.com/how-to-disable-links/#article-header-id-1) contained in `Disabled` would work? That seems to match with the spec for `a`:

> If the a element has no href attribute, then the element represents a placeholder for where a link might otherwise have been placed
> – [HTML Living Standard for `a`](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-a-element)

A quick test of `a` with no href in Safari + VoiceOver seems to work okay, but might not be very clear– there's no indication that the item is anything other than text, it does not communicate that "placeholder link" status. Some accessibility team feedback would probably be good here.

**Desktop (please complete the following information):**
 - OS: Mac
 - Browser: Chrome & Safari
 - Gutenberg: master @ 514e61298d5c056ef6ec4d3859b959dc46735dec

**Comment by afercia** [comment]
Thanks @ryelle for labelling this issue for accessibility feedback. Seems to me this is an interesting case. I'd like to split a few considerations in two groups:

## From a code perspective 

While the documentation of the `Disabled` component initially refers to "descendant tabbable elements", it makes then clear that this component is meant to be used for **form inputs**.

Looking at the code, this is exactly what it does:

https://github.com/WordPress/gutenberg/blob/0e46e9ef92d27aa4af849c482c81216b9c579410/packages/components/src/disabled/index.js#L66-L80

- adds a `disabled` attribute to all the form controls, see `DISABLED_ELIGIBLE_NODE_NAMES`
- if one of the focusable elements has a tabindex attribute, it removes it (I guess this is meant for non-natively focusable elements with a negative `-1` tabindex)
- removes `contenteditable` 

Basically the JS part doesn't do anything related to _links_. 

Actually, the Latest Comments content is not clickable because of the CSS part. It adds a CSS generated element that overlays all the block content, plus a couple `pointer-events: none` properties thus making the whole block not clickable.

At this point, I'm not sure why `Disabled` is used for Latest Comments to start with, when all that's needed is a few lines of CSS that could be added directly to the Latest Comments component.

Basically, seems to me `Disabled` is used incorrectly and its usage goes against its documentation.

Also, I'm not sure I understand why `Disabled` is used for the Latest Comments block and it's not for Latests Posts, Archives, etc. 

## Usability and Accessibility

The fact the block content can't be clicked with a pointing device but it can still be navigated through (and activated) when using a keyboard is definitely not ideal. I don't see a reason why the interaction with the wrapped component content should be different for users depending on the device / assistive technology they use.

Also, links can't be really "disabled". One option would be omitting the `href` attribute as @ryelle mentioned. However, I'm not sure interaction with links should be prevented in the first place.

These are links: they're supposed to behave like links also in a preview/edit context. Thinking also at the long term goal of inline-editing, I wouldn't recommend to try to prevent interaction with links.

**Comment by ryelle** [comment]
> While the documentation of the Disabled component initially refers to "descendant tabbable elements", it makes then clear that this component is meant to be used for form inputs.

I disagree, the form note comes in under "Usage" as an example. In my reading, "disables descendant tabbable elements and prevents pointer interaction" should include links, as they can be tabbed to. The fact that the JS doesn't currently handle this is why I opened this issue.

> I don't see a reason why the interaction with the wrapped component content should be different for users depending on the device / assistive technology they use.

Exactly– we have a component which removes interactivity from some interaction modes, but not all. 

Maybe there's a separate conversation to be had about whether `Disabled` should be used for these blocks, but this issue is mainly about the behavior of `<Disabled />`.

**Comment by afercia** [comment]
> I disagree, the form note comes in under "Usage" as an example.

Usually, documentation is written _after_ the code has been coded 🙂My point is only that the documentation seems inaccurate and doesn't reflect what the code does.

> In my reading, "disables descendant tabbable elements and prevents pointer interaction" should include links, as they can be tabbed to. 

- "disables descendant tabbable elements": actually the code doesn't disable them
- "pointer interaction" should be _pointing devices_ interaction, to be more correct

Links can't really be disabled. Without a `href` they're not links and they would be semantically incorrect. Also, I'm not sure how server-side rendered component could output two versions, one with links and one without links. Regardless, I don't think it should be done 🙂 

In my previous comment I meant the interaction should be the same with any device: allow tabbing and clicking on any element. I don't see any value in disabling native interaction.

Considering all this, I think the `Disabled` component _should not_ be used this way and probably should be considered for removal.

**Comment by ryelle** [comment]
>  I'm not sure how server-side rendered component could output two versions, one with links and one without links.

It wouldn't, I'm suggesting that the Disabled component should be taking care of that – just like it removes `tabindex` and sets `contenteditable`.

> I don't see any value in disabling native interaction.

Well, these aren't really editable links, so what would happen now if you click on a link: the page navigates away from the editor. You can't edit the content from ServerSideRender, so there's no benefit to being able to interact with it in the editor.

**Comment by kjellr** [comment]
In this case, the links are purely illustrative and not mean to be used for navigation. Since they're not clickable using a pointer device, I agree that you shouldn't be able to invoke them via the keyboard either. 

This seems relevant to #13771, where Gutenberg is currently being inconsistent in whether or not to entirely disable dropdowns in these blocks too. We [discussed this at a design triage](https://github.com/WordPress/gutenberg/pull/13783#discussion_r271385580), and decided that the dropdowns shouldn't be actionable. If that meant we needed to disable them, then that made the most sense there.

**Comment by afercia** [comment]
> you shouldn't be able to invoke them via the keyboard either.

I disagree 🙂 Users exploring the content with assistive technologies need to be able to read the content semantic and understand those are links. I'd strongly advice againt manipulating the content in the editing context.

Also considering the long term goal of inline editing, that seems to me a path that shouldn't be taken.

>  dropdowns shouldn't be actionable

I'd agree but those are meant for editing so it makes sense to disable them. Instead, when _exploring_ the content, users need to get the content with no alteration whatsoever.

**Comment by youknowriad** [comment]
Closing as resolved: the Disabled implementation now uses the native inert attribute, which removes contained links from keyboard interaction — fixing the original bug. The 2019 design debate about perceivability was effectively settled by the platform adopting inert as the standard. Thanks both for the thoughtful back-and-forth!
